### PR TITLE
Redirect replication requests when shard count is different

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Redirect replication requests when shard count is different

**Why?**
When replicate workflow between different shard count, it requires redirect to different shards.

**How did you test it?**
Unit tests

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/replication/task_executor.go [hunk 1]
```diff
@@ -174,7 +170,7 @@ func (e *taskExecutorImpl) handleActivityTask(
 	ctx, cancel := e.newTaskContext(ctx, attr.NamespaceId)
 	defer cancel()
 
-	err = e.historyEngine.SyncActivity(ctx, request)
+	_, err = e.shardContext.GetHistoryClient().SyncActivity(ctx, request)
```
Comment (meiliang86) [inline service/history/replication/task_executor.go:173]: Will there be additional cost when we always go through history client? Should we still go through history engine if the request is for the local shard?

Comment (wxing1292) [reply service/history/replication/task_executor.go:173]: we can certainly add additional wrapper, but the ROI is low?

Comment (wxing1292) [reply service/history/replication/task_executor.go:173]: actually, if we decide to do perf optimization, the optimization should be done within the wrapper of client

**Review by wxing1292 — APPROVED** [review]

File: service/history/replication/task_executor.go (hunk 1 above)
Comment (yux0) [reply service/history/replication/task_executor.go:173]: > actually, if we decide to do perf optimization, the optimization should be done within the wrapper of client

Yes. There will be additional network cost but I don't think it is a concern at this point. I can also add an additional check for now and move it to the wrapper of client.

**Review by meiliang86 — APPROVED** [review]

File: service/history/replication/task_executor.go (hunk 1 above)
Comment (meiliang86) [reply service/history/replication/task_executor.go:173]: Ok Let's leave it for now and address it in case it becomes an issue in prod.

**Review by yycptt — APPROVED** [review]

**Pull request merged**